### DISCUSSION TITLE
LOAD SHAPE AND OPENSTREETMAP CORRECTLY INTO MAPSUI MAP NET MAUI

### DISCUSSION THREAD

**Discussion by ghost** [Mapsui/Mapsui#2667] [Development]
Hello everyone...

I'm going crazy!!

Let me explain... I'm starting with Net MAUI and, at the same time, with MAPSUI.

The fact is that everything is fine... I create the UI, I see OpenStreetMap, etc... but the mouse coordinates widget indicates things like -985479, 5296763. Strange coordinates where they exist, at least for my area.
With the positioning data from the mobile's GPS it correctly tells me where I am.
The problem comes when I try to load a SHAPE file.

When creating the map I indicate CRS EPSG:25829 of my area

mapa.CRS = "EPSG:25829"

... and the SHAPE file comes with the same CRS from QGIS

var shape = new ShapeFile(miShape, true); //EPSG:25829

, but when loading it it comes out about 2000 km displaced to the east. But interestingly their coordinates are correct, so I understand that it is the OpenStreetMap Tile that is offset to the west.

I also load a MyLocationLayer, check the GPS location data and it locates me correctly on OpenStreetMap.

I can't figure out why this is happening, but I'm going crazy trying to figure out how to get it to load everything in its correct location.

Any help that can guide me?

Thank you very much in advance.

**Comment by pauldendulk** [comment]
For EPSG:25829 you need DotSpatiaProjection. Look at the ShapeFileProjectionSample (category "Projection") for how to use that in combination with shp. The ProjectingProvider needs to have CRS = "EPSG:3857" and it's source shp CRS = "EPSG:25829".

**Reply by pauldendulk** [reply]
You might want to start with trying to project a single point from EPSG:25829 to EPSG:3857.

**Reply by ghost** [reply]
OK thanks. I have seen it and I have tried to follow the example but I have not succeeded, I have surely done something wrong. I'll try again to see if I have more luck. thank you so much. all the best

**Comment by ghost** [comment]
Hello again...

You'll have to forgive my incompetence, but trying to follow the example to reproject a SHAPE file, I find that I can't get out of this error:

var dotSpatialProjection = new DotSpatialProjection(); (namespace don't find)

DotSpatialProjection not recognized

I downloaded what I thought was necessary, a package called DotSpatial Projection (nuget.org) but it doesn't recognize it either.

In short... I'm still lost on how to reproject my SHAPEs so that they render correctly on OpenStreetMap and BigMaps.

**Reply by charlenni** [reply]
Did you add Mapsui.Extensions from nuget.org? It contains the class.

**Reply by ghost** [reply]
Thank you all very much... I have managed to reproject my SHAPEFILES, it seems that I did not have MAPSUI.extensions loaded.

Now my question... when loading the shape with new ShapeFile(shapePath, true, true, dotprojection), it always returns CRS = NULL but one of the options is to read the PRJ and I have set it to true. I know, because I have prepared the SHAPE, which is in EPSG:25829, but the normal thing is that I do not know a priori the coordinate system of the shape since it will depend on the user. 

Right now, as I know, I assign the CRS, but... is there a way to know which CRS the shape originates from so that I can assign it and have it perform the reprojection correctly?

Thank you so much

**Reply by pauldendulk** [reply]
Yes, you can create the ShapeFile provider with DotSpatialProjection as argument and it will use to *.prj to determine the CRS of the provider.

```csharp
    [Test]
    public void ShapeFileReadTest()
    {
        // arrange
        var directory = Path.GetDirectoryName(typeof(ProjectionTests).Assembly.Location);
        var countriesPath = Path.Combine(directory!, "GeoData", "World", "countries.shp");

        // act
        using var shapeFile = new ShapeFile(countriesPath, false, true, new DotSpatialProjection());

        // assert
        Assert.AreEqual(shapeFile.CRS, "EPSG:4326");
    }
```
